### PR TITLE
TensorView fixes.

### PR DESCRIPTION
**Pull request by mzient**
Fix broken assert.
Add convert_dim function to/from static/dynamic.
Add type-changing constructors.

Signed-off-by: Michal Zientkiewicz <michalz@nvidia.com>

### CODE REVIEW

File: dali/kernels/tensor_shape.h [hunk 1]
```diff
@@ -381,7 +381,9 @@ struct TensorListShapeBase {
   TensorShape<tensor_ndim> tensor_shape(int64_t sample) const {
     static_assert(tensor_ndim == sample_ndim || sample_ndim == DynamicDimensions
                   || tensor_ndim == DynamicDimensions, "Cannot convert to other static ndim");
-    assert(tensor_ndim == sample_dim() && "Cannot convert to other ndim");
+    if (tensor_ndim != DynamicDimensions && sample_ndim != DynamicDimensions) {
```
Comment (klecki) [inline dali/kernels/tensor_shape.h]:
```suggestion
    if (tensor_ndim != DynamicDimensions) {
```
Even if sample_ndim is dynamic you still want to check if static value of tensor_ndim matches current sample_dim().

File: dali/kernels/tensor_shape.h [hunk 2]
```diff
@@ -645,6 +647,35 @@ bool is_uniform(const TensorListShape<ndim> &tls) {
   return true;
 }
 
+template <int out_dim, int in_dim>
+TensorShape<out_dim> convert_dim(const TensorShape<in_dim> &in) {
```
Comment (klecki) [inline dali/kernels/tensor_shape.h]: This is a duplicate of those below?

File: dali/kernels/tensor_shape.h [hunk 3]
```diff
@@ -645,6 +647,35 @@ bool is_uniform(const TensorListShape<ndim> &tls) {
...
+  TensorShape<out_dim> out;
+  out.resize(in.size());
+  for (int i = 0; i < out.size(); i++)
+    out[i] = in[i];
+  return out;
+}
+
+template <int out_dim, int in_dim>
+typename std::enable_if<(out_dim != DynamicDimensions), TensorListShape<out_dim>>::type
+convert_dim(const TensorListShape<in_dim> &in) {
```
Comment (klecki) [inline dali/kernels/tensor_shape.h]: As we have a convert_dim as global, maybe to_static should be adjusted as well? To have some consistency.

File: dali/kernels/tensor_shape.h [hunk 4]
```diff
@@ -645,6 +647,35 @@ bool is_uniform(const TensorListShape<ndim> &tls) {
...
+  static_assert(out_dim == DynamicDimensions || in_dim == DynamicDimensions ||
+                in_dim == out_dim, "Incompatible number of dimensions"
+                " - must be equal or at least one side must be dynamic");
+  TensorListShape<out_dim> out = in.template to_static<out_dim>();
+  return out;
+}
+
+template <int out_dim, int in_dim>
+typename std::enable_if<(out_dim == DynamicDimensions), TensorListShape<out_dim>>::type
+convert_dim(const TensorListShape<in_dim> &in) {
```
Comment (klecki) [inline dali/kernels/tensor_shape.h]: `TensorListShape<in_dim> &&in` variant?

Comment (mzient) [reply dali/kernels/tensor_shape.h]: I see little reason to have it. In most cases, we need to copy anyway. The only scenario is that inside some template we have a no-op conversion from dynamic to dynamic - but even then, I don't think a move is likely.

Comment (mzient) [reply dali/kernels/tensor_shape.h]: Second thought: for Tensor**List**Shape it might be useful.

File: dali/kernels/tensor_shape.h [hunk 5]
```diff
@@ -645,6 +647,52 @@ bool is_uniform(const TensorListShape<ndim> &tls) {
...
+}
+
+
+template <int out_dim, int in_dim>
+typename std::enable_if<(out_dim != DynamicDimensions), TensorListShape<out_dim>>::type
+convert_dim(TensorListShape<in_dim> &&in) {
+  static_assert(out_dim == DynamicDimensions || in_dim == DynamicDimensions ||
+                in_dim == out_dim, "Incompatible number of dimensions"
+                " - must be equal or at least one side must be dynamic");
+  TensorListShape<out_dim> out = std::move(in).template to_static<out_dim>();
```
Comment (klecki) [inline dali/kernels/tensor_shape.h]: :+1:

File: dali/kernels/tensor_view.h [hunk 6]
```diff
@@ -267,12 +316,16 @@ struct TensorListView<Backend, DataType, DynamicDimensions>
...
 
-  template <int other_sample_ndim>
-  TensorListView(const TensorListView<Backend, DataType, other_sample_ndim> &other)
-      : Base(other.data, other.shape, other.offsets) {}
-  template <int other_sample_ndim>
-  TensorListView(TensorListView<Backend, DataType, other_sample_ndim> &&other)
-      : Base(other.data, std::move(other.shape), std::move(other.offsets)) {
+  template <int other_sample_ndim, typename U>
+  TensorListView(const TensorListView<Backend, U, other_sample_ndim> &other)
+  : Base(other.data, other.shape, other.offsets) {
```
Comment (klecki) [inline dali/kernels/tensor_view.h]: This is one of three ways to indent initializer list that I can (sadly) find in our code base, it is not compatible with style-guide https://google.github.io/styleguide/cppguide.html#Constructor_Initializer_Lists and .clang-format. 
And it is inconsistent in this file.

I would suggest to stick to one thing with formatting, and advise to use clang-format for the new files.

**Pull request merged**